aten/src/ATen/native/Unique.cpp

```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+// which made this part almost a free launch.
+//
+// This kernel also implements a `equal_nan` flag which has same
+// function as NumPy's unique. Currently this is always disabled.
+//
+template <typename scalar_t>
+std::tuple<Tensor, Tensor, Tensor> unique_cpu_sorted_template(
+    const Tensor& self,
+    const bool return_inverse,
+    const bool return_counts,
+    const bool equal_nan = false) {
+  const Tensor& input = self.contiguous();
+
+  int64_t numel = input.numel();
+  Tensor output = at::empty({0}, self.options());
+  Tensor inverse_indices = at::empty({0}, self.options().dtype(kLong));
+  Tensor counts = at::empty({0}, self.options().dtype(kLong));
+
+  if (numel == 0) {
+    if (return_inverse) {
+      inverse_indices.resize_(input.sizes());
+    }
+    return std::make_tuple(output, inverse_indices, counts);
+  }
+
+  // index of first unique in each consecutive section
+  // this is used to compute counts for parallelization purpose
+  Tensor unique_index = at::empty({0}, self.options().dtype(kLong));
+
+  // original behavior with unique on scalar tensor
+  // is to return a output size of ([1]), `flatten` here will do the job
+  auto input_flattened = input.flatten();
+
+  Tensor input_sorted, indices;
+  std::tie(input_sorted, indices) = input_flattened.sort();
+
+  scalar_t* input_sorted_data = input_sorted.data_ptr<scalar_t>();
+  int64_t* indices_data = indices.data_ptr<int64_t>();
+
+  // `mask` keeps track of whether it is the first unique
+  // in the sorted input sequence
+  Tensor mask = at::empty({numel}, self.options().dtype(kBool));
+  auto mask_acc = mask.accessor<bool, 1>();
+
+  int num_threads = at::get_num_threads();
+  std::vector<int64_t> unique_count_thread(num_threads, 0);
+  std::vector<int64_t> offset_thread(num_threads, 0);
+
+  // the first element is always true
+  mask_acc[0] = true;
+
+  // we can parallel on [1, numel) but we need to make sure it has
+  // the same parallel scope with the next loop
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    if (begin == 0) { begin += 1; }
+    for (const auto i : c10::irange(begin, end)) {
+      mask_acc[i] = input_sorted_data[i] != input_sorted_data[i - 1];
+    }
+  });
+
+  // handle equal the NaN's
+  bool last_element_isnan = _isnan<scalar_t>(input_sorted_data[numel - 1]);
+  if (input.is_floating_point() && equal_nan && last_element_isnan) {
+    auto is_nan = [](const scalar_t& val){ return _isnan<scalar_t>(val); };
+    auto firstnan = std::find_if(input_sorted_data, input_sorted_data + numel, is_nan);
+    int64_t firstnan_index = std::distance(input_sorted_data, firstnan);
+
+    mask_acc[firstnan_index] = true;
+    if (firstnan_index + 1 < numel) {
+      for (const auto i : c10::irange(firstnan_index + 1, numel)) {
+        mask_acc[i] = false;
+      }
+    }
+  }
+
+  // calculate unique count from each thread
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    int tid = at::get_thread_num();
```

Comment:
It is always beneficial to set grain_size to zero to force parallelization?

Reply:
No. For a small problem size, parallel it would be slower. But in practice, the difference of setting grain size or not does not matter very much. Because if the problem size is very small, then it would only take very small amount of time.

Usually for simple operators we set grain_size to be `at::internal::grain_size` which is 32K and for complex operators just set grain_size to 0. And the parallel loop in this patch has been changed to use the default grain size.